### DISCUSSION TITLE
BaseView 프로토콜 도입 여부

### DISCUSSION THREAD

**Discussion by EnnzoKim** [Favor-Gift-Reminder/Favor-iOS#16] [Show and tell]
```swift
/// UIView를 상속하는 Custom View를 구현할 때 필요한 프로토콜입니다
/// 
/// 명시적으로 init 메서드에 아래 세 가지 메서드를 호출해야합니다.
protocol BaseView {
  
  /// View와 관련된 Style을 설정합니다.
  ///
  /// ```
  /// func setupStyles() {
  ///   navigationController?.navigationBar.tintColor = .white
  ///   view.backgroundColor = .white
  ///   // codes..
  /// }
  /// ```
  func setupStyles()
  
  /// UI 프로퍼티를 view에 할당합니다.
  ///
  /// ```
  /// func setupLayouts() {
  ///   self.view.addSubview(label)
  ///   self.stackView.addArrangedSubview(label)
  ///   self.label.layer.addSubLayer(gradientLayer)
  ///   // codes..
  /// }
  /// ```
  func setupLayouts()
  
  /// UI 프로퍼티의 제약조건을 설정합니다.
  ///
  /// ```
  /// func setupConstraints() {
  ///   // with SnapKit
  ///   label.snp.makeConstraints { make in
  ///     make.edges.equalToSuperview()
  ///   }
  ///   // codes..
  /// }
  /// ```
  func setupConstraints()
}
```

현재 `BaseViewController`라는 상속의 이점을 통해서 
UIViewController의 UI 레이아웃 및 제약조건의 컨벤션을 맞춰주고 있는데,

정작 `UIView`를 상속하는 Custom View를 구현할 떄는 마땅한 옵젝이 없어서 한번 구현해봤어
`UIView`를 상속하는 CustomView는 모양새가 다양하기 떄문에 
(`UIImageView`, `UICollectionViewCell`, `UILabel` 등등 ...) 

상속을 사용할 수는 없었고, 이 이유떄문에 프로토콜로 구현했어
프로토콜로 구현했기 떄문에 위 세 가지 메서드는 `init` 메서드에서 명시적으로 불러줘야 할거야 
이 구현 내용은 현재 `Favor15 OnboardingVC-UI` 브랜치에만 존재하고 
이것을 도입할지 안 할지 의견이 궁금해서 Discussion 열어봤어

한번 보고 의견 부탁해!

**Comment by nomatterjun** [comment]
좋아좋아
안그래도 커스텀뷰 만들면서 뷰컨처럼 UI 셋업들 프로토콜이나 상속할 코드가 있으면 좋겠다고 생각하고 있었어

dev 브랜치에 바로 커밋하고 브랜치에 마지시켜서 작업하면 좋을 것 같아!

👍👍

**Reply by EnnzoKim** [reply]
dev에서 커밋하고 푸쉬해도 될까?
충돌 걱정이 좀 있ㄴ ㅔ.. 별 상관 없으려나 ..?

**Reply by nomatterjun** [reply]
응응 변경사항이 많은 것도 아니고
있어도 해결하면 되는거니까!

**Comment by EnnzoKim** [comment]
`dev`에 적용 완료 !
